src/content/docs/en/guides/incremental-static-regeneration.mdx

```diff
@@ -0,0 +1,276 @@
+ - `Cache-Control: s-maxage=60` tells the caching server to cache the response for 60 seconds. You probably want to use this for most cases.
+ - `Cache-Control: max-age=60` tells both browsers and servers to cache the response for 60 seconds. You may use this when your content is unlikely to be edited. If a page with `max-age` updates after it has been visited by a user, their browser will show them the cached version until the age of the content is past `max-age`.
+ - `Cache-Control: private, max-age=60` tells the browser to cache the response for 60 seconds, but tells the servers to not cache it. This is useful when you want to provide quick navigation, and the content includes details specific to the visitor.
+
+#### About the Web Cache API
+Deno and Cloudflare implement the Web Cache API, a way to store and retrieve `Response` objects using javascript.
+To use this API, call the `open` method on the globally available `caches` object with a name of your choosing.
+```ts
+const cache = await caches.open("astro")
+```
+You just opened a [`Cache`](https://developer.mozilla.org/en-US/docs/Web/API/Cache)! You can now call `await cache.put(request, response)` to save a response, and `await cache.match(request)` to retrieve it. See [web.dev's Cache API quick guide](https://web.dev/cache-api-quick-guide/) if you want to learn more.
+
+### Self-hosted Node.js
+ISR can be implemented in Node.js by writing a caching middleware. The following middleware caches every response in memory for 60 seconds:
+
+```ts title="src/middleware.ts"
+import { defineMiddleware } from "astro:middleware"
+
+const cache = new Map
+
+export const onRequest = defineMiddleware(async ({ request }, next) => {
+    
+    let expiresIn = 60
+    
+    const cached = cache.get(request.url)
+    
+    if (cached) {
+        if (cached.expires > Date.now()) {
+            return cached.response.clone()
+        }
+        else {
+            cache.delete(request.url)
+        }
+    }
+    
+    const response = await next()
+    
+    cache.set(request.url, {
+        expires: Date.now() + expiresIn * 1000,
+        response: response.clone()
+    })
+    
+    return response
+})
+```
+:::tip
+It is a good idea to skip caching when the response code is 500 (Internal Server Error).
+:::
+
+#### Advanced: Per-page revalidation time
+You may want some pages to be always rendered fresh, while others to be kept saved for hours. Astro makes this possible with middleware locals.
+
+You can create a function that changes the `expiresIn` variable, and add it to locals. Now each astro page can decide how long it should be cached for!
+```ts title="src/middleware.ts" ins={9}
+import { defineMiddleware } from "astro:middleware"
+
+const cache = new Map
+
+export const onRequest = defineMiddleware(async ({ request }, next) => {
+    
+    let expiresIn = 60
+    
+    locals.revalidate = seconds => { expiresIn = seconds }

```

Comment:
I don't think `locals` is globally exposed, is it part of `onRequest` ctx argument?
```suggestion
export const onRequest = defineMiddleware(async ({ request, locals }, next) => {
    
    let expiresIn = 60
    
    locals.revalidate = seconds => { expiresIn = seconds }

```

Reply:
You're exactly right! Thanks